test/metabase/models/setting_test.clj

```diff
@@ -1204,45 +1204,189 @@
               #"^Surprise!$"
               (#'setting/realize x)))))))
 
-(deftest valid-json-setting-test
-  (mt/with-temp-env-var-value ["MB_TEST_JSON_SETTING" "[1, 2]"]
-    (is (nil? (setting/validate-settings-formatting!)))))
-
-(defn- get-parse-exception [raw-value]
-  (mt/with-temp-env-var-value ["MB_TEST_JSON_SETTING" raw-value]
-    (try
-      (setting/validate-settings-formatting!)
-      (throw (java.lang.RuntimeException. "This code should never be reached."))
-      (catch java.lang.Exception e e))))
-
-(defn- assert-parser-exception! [ex cause-message]
-  (is (= "Invalid JSON configuration for setting: test-json-setting" (ex-message ex)))
+
+(defmacro define-setting-for-type [format]
+  `(defsetting ~(symbol (str "test-" (name format) "-setting"))
+     "Setting to test validation of this format - this only shows up in dev"
+     :type ~(keyword (name format))))
+
+(defmacro define-settings-for-types [& formats]
+  (list 'dorun (cons 'list (for [f formats] `(define-setting-for-type ~f)))))
+
+;; Create stub settings for the remaining types that weren't used by any previous tests
+(define-settings-for-types :double :keyword :integer :positive-integer :timestamp)
```

Comment:
I couldn't think of a quick way to catch if a new type is added without updating this field. I could open an issue for doing that? This rabbit hole keeps going 🙂

Reply:
As we discussed I think this hits the limit of where we want to go for effort vs reward with regards to future-proofing. If a new setting type is created but we forget to handle custom validation logic.

Anyway, I think there's a simpler pattern than the `define-settings-for-types` macro.

```
(doseq [type setting-types]
  (define-setting-for-type type))
```

 Thankfully clojure is an imperative language too :)

Reply:
It might be a good exercise to think about how to enforce this constraint if we wanted though. I think a check on executing a defsetting macro would do it, similar to how you're doing it in [this PR](https://github.com/metabase/metabase/pull/37624/files#diff-1fb7cd6dc14f96149b2b775f598268d8053ee68dca276c40001b41ca4965cff6), but the implementation could be done in a few ways.

Reply:
Found a great way to enumerate them: `(set (keys (methods get-value-of-type)))`

Reply:
> Anyway, I think there's a simpler pattern than the define-settings-for-types macro.

I don't think that doseq works, because the macro will only see the symbol `type, or am I missing something?

Reply:
I decided there wasn't much point in making this a one liner, and actually it's quite nice to use the definitions as "headers" for each set of tests. 

I added a test at the end to validate that we haven't missed any sections (i.e. we've defined all the header vars). 

I also turned the method which we removed the logging from into a multimethod so that we could also validate its completeness.